### PR TITLE
Add initial implementation of the unprefixed fullscreen API

### PR DESCRIPTION
**Pull request by cdumez**
#### 6331c0ff7a9324e34aac07a9745b7f0848ea1cc5
<pre>
Add initial implementation of the unprefixed fullscreen API
<a href="https://bugs.webkit.org/show_bug.cgi?id=246103">https://bugs.webkit.org/show_bug.cgi?id=246103</a>

Reviewed by Tim Nguyen.

Add initial implementation of the unprefixed fullscreen API:
- <a href="https://fullscreen.spec.whatwg.org/#api">https://fullscreen.spec.whatwg.org/#api</a>

The feature is behind an experimental feature flag, off by default.

This is needed to unblock the WPT testing of the screen orientation API since
screen orientation locking requires being in fullscreen.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/capability-delegation/delegate-fullscreen-request-popup-same-origin.https.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/capability-delegation/delegate-fullscreen-request-subframe-same-origin.https.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/idlharness.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/active-lock-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/event-before-promise-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/lock-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/lock-sandboxed-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/lock-unlock-check-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/onchange-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/onchange-event-subframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/screen-orientation/orientation-reading-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/dom/Element+Fullscreen.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::dispatchFullscreenChangeEvents):
* Source/WebCore/html/HTMLAttributeNames.in:

Canonical link: <a href="https://commits.webkit.org/255317@main">https://commits.webkit.org/255317@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e5eb0208d121e25a72a80d4de69570775ffaa92e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92145 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101930 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1374 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29766 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84586 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98099 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/879 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78668 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27820 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82798 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82391 "Found 1 new API test failure: TestWebKitAPI.Fullscreen.LayoutConstraints (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70857 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36194 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16404 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33950 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17522 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3686 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37821 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36649 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/dom/Document+Fullscreen.idl [hunk 1]
```diff
@@ -43,4 +43,11 @@
 
     [NotEnumerable] attribute EventHandler onwebkitfullscreenchange;
     [NotEnumerable] attribute EventHandler onwebkitfullscreenerror;
+
+    // Standard version.
```
Comment (weinig) [inline Source/WebCore/dom/Document+Fullscreen.idl]: Give this is the standard one, I think you should swap things by putting this at the top of the interface, and annotating the non-standard ones as non-standard with the comments we usually add for such things.

File: Source/WebCore/dom/DocumentFullscreen.h [hunk 2]
```diff
@@ -45,8 +47,23 @@ class DocumentFullscreen {
     static bool webkitFullScreenKeyboardInputAllowed(Document& document) { return document.fullscreenManager().isFullscreenKeyboardInputAllowed(); }
     static Element* webkitCurrentFullScreenElement(Document& document) { return document.ancestorElementInThisScope(document.fullscreenManager().currentFullscreenElement()); }
     static void webkitCancelFullScreen(Document& document) { document.fullscreenManager().cancelFullscreen(); }
+    static void exitFullscreen(Document&, Ref<DeferredPromise>&&);
 };
 
+// FIXME: Add proper support.
+// https://fullscreen.spec.whatwg.org/#exit-fullscreen
+inline void DocumentFullscreen::exitFullscreen(Document& document, Ref<DeferredPromise>&& promise)
```
Comment (weinig) [inline Source/WebCore/dom/DocumentFullscreen.h]: Is this really a win to be inline?

File: Source/WebCore/dom/DocumentOrShadowRoot.idl [hunk 3]
```diff
@@ -44,4 +44,7 @@ interface mixin DocumentOrShadowRoot {
     Element? elementFromPoint(double x, double y);
     sequence<Element> elementsFromPoint(double x, double y);
     // CaretPosition? caretPositionFromPoint(double x, double y); // FIXME: Implement this.
+
+    // https://fullscreen.spec.whatwg.org/#api
+    [LegacyLenientSetter, Conditional=FULLSCREEN_API, EnabledBySetting=UnprefixedFullscreenAPIEnabled] readonly attribute Element? fullscreenElement;
```
Comment (weinig) [inline Source/WebCore/dom/DocumentOrShadowRoot.idl]: This should be in its own file as a `partial interface mixin`

File: Source/WebCore/dom/DocumentFullscreen.h (hunk 2 above)
Comment (cdumez) [reply Source/WebCore/dom/DocumentFullscreen.h]: Will move to a new cpp file.

File: Source/WebCore/dom/Document+Fullscreen.idl (hunk 1 above)
Comment (cdumez) [reply Source/WebCore/dom/Document+Fullscreen.idl]: Makes sense, will do.

File: Source/WebCore/dom/DocumentOrShadowRoot.idl (hunk 3 above)
Comment (cdumez) [reply Source/WebCore/dom/DocumentOrShadowRoot.idl]: Ok.

File: Source/JavaScriptCore/runtime/JSFunction.h [hunk 4]
```diff
@@ -27,6 +27,7 @@
 #include "InternalFunction.h"
 #include "JSCallee.h"
 #include "JSScope.h"
+#include "VM.h"
```
Comment (nt1m) [inline Source/JavaScriptCore/runtime/JSFunction.h]: nit: seems unrelated.

File: LayoutTests/imported/w3c/web-platform-tests/dom/idlharness.window-expected.txt [hunk 5]
```diff
@@ -342,7 +342,7 @@ PASS Document interface object length
 PASS Document interface object name
 PASS Document interface: existence and properties of interface prototype object
 PASS Document interface: existence and properties of interface prototype object's "constructor" property
-PASS Document interface: existence and properties of interface prototype object's @@unscopables property
+FAIL Document interface: existence and properties of interface prototype object's @@unscopables property assert_not_equals: Document.prototype[Symbol.unscopables] has unexpected property "fullscreen" got disallowed value -1
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/dom/idlharness.window-expected.txt:345]: what is this about?

File: Source/WebCore/dom/FullscreenManager.cpp [hunk 6]
```diff
@@ -78,25 +82,32 @@ void FullscreenManager::requestFullscreenForElement(Ref<Element>&& element, Full
...
+        }
+        // We do not allow pressing the Escape key as a user gesture to enter fullscreen since this is the key
+        // to exit fullscreen.
+        if (UserGestureIndicator::currentUserGesture()->gestureType() == UserGestureType::EscapeKey) {
+            ERROR_LOG(LOGIDENTIFIER, "Current gesture is EscapeKey; failing.");
+            document().addConsoleMessage(MessageSource::Security, MessageLevel::Error, "The Escape key may not be used as a user gesture to enter fullscreen"_s);
+            failedPreflights(WTFMove(element), WTFMove(promise));
+            return;
+        }
     }
```
Comment (nt1m) [inline Source/WebCore/dom/FullscreenManager.cpp]: I prefer to not have if/else branches like this, I think all preflights checks should be done, unconditionally, for both APIs.

I do want to rework the preflights checks soon enough, so I'd rather avoid having to deal with multiple branches.

I don't think there is some web impact to add the `if (!document().domWindow() || !document().domWindow()->hasTransientActivation())` check to the old API, so I think we should be safe to go.

File: Source/WebCore/dom/Element.cpp [hunk 7]
```diff
@@ -4091,9 +4093,20 @@ static Element* parentCrossingFrameBoundaries(const Element* element)
...
+// https://fullscreen.spec.whatwg.org/#dom-element-requestfullscreen
+// FIXME: Add proper support.
+void Element::requestFullscreen(FullscreenOptions&&, Ref<DeferredPromise>&& promise)
+{
+    if (!document().isFullyActive()) {
+        promise->reject(Exception { TypeError, "Document is not fully active"_s });
+        return;
+    }
+    webkitRequestFullscreen(WTFMove(promise));
 }
```
Comment (nt1m) [inline Source/WebCore/dom/Element.cpp]: I'd like `webkitRequestFullscreen` to be a wrapper around `requestFullscreen`, not the other way round. 

```
    if (!document().isFullyActive()) {
         promise->reject(Exception { TypeError, "Document is not fully active"_s });
         return;
     }
```
can move into `FullscreenManager::requestFullscreenForElement()` as a preflight check (for both APIs).

Another benefit is that `webkitRequestFullscreen` can avoid taking in the `promise` as parameter as well.

File: Source/WebCore/dom/DocumentFullscreen.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        promise->resolve();
+    });
+}
+
+// https://fullscreen.spec.whatwg.org/#dom-document-fullscreenenabled
+bool DocumentFullscreen::fullscreenEnabled(Document& document)
+{
+    if (!document.isFullyActive())
+        return false;
+    return document.fullscreenManager().isFullscreenEnabled();
```
Comment (nt1m) [inline Source/WebCore/dom/DocumentFullscreen.cpp:55]: Would be totally reasonable to move `document.isFullyActive()` into `document.fullscreenManager().isFullscreenEnabled()` IMO.

File: Source/WebCore/dom/DocumentFullscreen.cpp [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+namespace WebCore {
+
+// FIXME: Add proper support.
+// https://fullscreen.spec.whatwg.org/#exit-fullscreen
+void DocumentFullscreen::exitFullscreen(Document& document, Ref<DeferredPromise>&& promise)
+{
+    if (!document.isFullyActive() || !document.fullscreenManager().fullscreenElement()) {
+        promise->reject(Exception { TypeError, "Not in fullscreen"_s });
+        return;
+    }
```
Comment (nt1m) [inline Source/WebCore/dom/DocumentFullscreen.cpp:43]: nit: I think we should strive to have as few differences with the prefixed version as possible (so aliasing becomes easier later on).

The `!document.isFullyActive()` check is something that can be moved inside `FullscreenManager` and there is already a check similar to `!document.fullscreenManager().fullscreenElement()` inside `FullscreenManager` (the check for the empty fullscreen stack).

If you're worried about potential fallout, I would rather catch it now than later, and I can adjust accordingly later.

**Review by nt1m — COMMENTED** [review]
Looks generally fine to me, though I'd like less differences between the prefixed and unprefixed APIs, it seems to be mostly about preflight checks at the moment, which I think are reasonable to change in the old API as well.

**Comment by cdumez** [comment]
> Looks generally fine to me, though I'd like less differences between the prefixed and unprefixed APIs, it seems to be mostly about preflight checks at the moment, which I think are reasonable to change in the old API as well.

I really do not want to change the behavior of the shipping API in this large patch. I think this carries risk and such large patches for feature A should not change the behavior of feature B.

File: Source/WebCore/dom/DocumentFullscreen.cpp (hunk 9 above)
Comment (cdumez) [reply Source/WebCore/dom/DocumentFullscreen.cpp:43]: Same, I think this is a good idea but not in this patch. We can try and harmonize the 2 code paths in a small followup patch. This way, if it introduces a regression, it is very obvious what the issue is.

**Comment by nt1m** [comment]
> > Looks generally fine to me, though I'd like less differences between the prefixed and unprefixed APIs, it seems to be mostly about preflight checks at the moment, which I think are reasonable to change in the old API as well.
> 
> I really do not want to change the behavior of the shipping API in this large patch. I think this carries risk and such large patches for feature A should not change the behavior of feature B.

@cdumez That's reasonable. Can we move all the new logic into FullscreenManager still and make `webkitRequestFullscreen` APIs wrap around `requestFullscreen`? (even if it means adding if/else checks).

I do plan to unify the code for both prefixed/unprefixed APIs anyway, so it would be easier if stuff wasn't scattered around.

**Comment by cdumez** [comment]
> > > Looks generally fine to me, though I'd like less differences between the prefixed and unprefixed APIs, it seems to be mostly about preflight checks at the moment, which I think are reasonable to change in the old API as well.
> > 
> > 
> > I really do not want to change the behavior of the shipping API in this large patch. I think this carries risk and such large patches for feature A should not change the behavior of feature B.
> 
> @cdumez That's reasonable. Can we move all the new logic into FullscreenManager still and make `webkitRequestFullscreen` APIs wrap around `requestFullscreen`? (even if it means adding if/else checks).

Yes, I'll give this a shot.

**Review by nt1m — APPROVED** [review]
LGTM, aside from my two comments (the unrelated header file change, and the regression in the test file).

File: Source/JavaScriptCore/runtime/JSFunction.h (hunk 4 above)
Comment (cdumez) [reply Source/JavaScriptCore/runtime/JSFunction.h]: Yes, was trying to fix the Windows build iirc. Looks like Windows EWS is still red though.
